### PR TITLE
Feature/jax builder class

### LINKED ISSUE #426: [FEATURE] General system builder for Jax-based systems
### Feature
A builder class for Jax-based Mava systems.

### Proposal
The builder will take as input a list of system components, which are themselves callbacks and build the system using a set of functions consisting of hooks. Functions will include building tables, datasets, adders, variable servers and clients, executor, trainers, evaluators and launching. 

### Testing
All functions should be tested with proper handling of hooks. It is important to handle single instance hooks as well as make sure that calls run when expected and only when expected using the Mixin.  

### Definition of done
A precise outline for the feature to be considered fully implemented, i.e. done/complete.

### Mandatory checklist before making a PR
* [ ] The success criteria laid down in “Definition of done” are met.
* [ ] Code is documented - docstrings for methods and classes, static types for arguments.
* [ ] Code is tested - unit, integration and/or functional tests are added.
* [ ] Documentation is updated - README, CONTRIBUTING, or other documentation.
* [ ] All functional tests are green.

### PR DESCRIPTION
**Pull request by arnupretorius**
## What?
General builder class for Jax systems.
## Why?
For building systems in a highly flexible way using callback system components.
## How?
Create a builder class with general-purpose builder methods and hooks.

closes #426 #443 #440

### CODE REVIEW

File: mava/callbacks/hook_mixin.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+class CallbackHookMixin(ABC):
+
+    ######################
+    # system builder hooks
+    ######################
+
+    callbacks: List
+
+    # INITIALISATION
+    def on_building_init_start(self) -> None:
```
Comment (AsadJeewa) [inline mava/callbacks/hook_mixin.py:31]: There is a very large number of hooks. Do we handle this as we may at some point need to inject code at any of these points and in that regard, we can add or remove hooks as necessary?

File: mava/systems/jax/builder.py [hunk 2]
```diff
@@ -13,28 +13,188 @@
...
+class Builder(SystemBuilder, CallbackHookMixin):
     def __init__(
         self,
         components: List[Any],
     ) -> None:
         """System building init
 
         Args:
-            components: system callback components
+            components: system callback blocks
```
Comment (AsadJeewa) [inline mava/systems/jax/builder.py:35]: so blocks are analogous to components that make up a system? why do we need the blocks class exactly?

File: mava/systems/jax/builder_test.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    def on_building_executor_environment(self, builder: SystemBuilder) -> None:
+        """_summary_"""
+        builder.blocks.env = (
+            builder.blocks.exec_logger,
+            self.config.executor_param_2,
+        )
+
+    def on_building_executor_environment_loop(self, builder: SystemBuilder) -> None:
+        """_summary_"""
+        if builder._executor_id != "evaluator":
```
Comment (AsadJeewa) [inline mava/systems/jax/builder_test.py:191]: Nice and clean :smile:

**Review by AsadJeewa — DISMISSED** [review]
Lots of code but everything looks good from my side

File: mava/systems/jax/system_test.py [hunk 4]
```diff
@@ -48,14 +47,18 @@ def dummy_str_plus_bool(self) -> None:
 
 
 # Mock builder
-class MockBuilder(Builder, MockCallbackHookMixin):
-    def __init__(self, components: List[Any]) -> None:
-        """Init for mock builder.
+class MockBuilder(MockCallbackHookMixin):
```
Comment (AsadJeewa) [inline mava/systems/jax/system_test.py:50]: isn't the point to test the actual mixin class so why are we creating a mock class?

File: mava/callbacks/hook_mixin.py (hunk 1 above)
Comment (arnupretorius) [reply mava/callbacks/hook_mixin.py:31]: Yes, this is exactly right, so in the future, we might add more hooks if needed, but for now, we are creating the ones we think are most necessary.

File: mava/systems/jax/builder.py (hunk 2 above)
Comment (arnupretorius) [reply mava/systems/jax/builder.py:35]: Oops I think this is a type that creeped in with a global find and replace 😅

Comment (arnupretorius) [reply mava/systems/jax/builder.py:35]: Will fix it.

File: mava/systems/jax/system_test.py (hunk 4 above)
Comment (arnupretorius) [reply mava/systems/jax/system_test.py:50]: This was before we had the actual mixin class but I still think it is good to create mock versions to isolate the testing to just the system. But that said, we must write some integration tests for all the system processes.

**Comment by arnupretorius** [comment]
Closing this PR since it is being subsumed here: #457.

**Pull request closed without merging**